lib/protobuf/nats/client.rb

```diff
@@ -137,34 +198,47 @@ def nats_request_with_two_responses(subject, data, opts)
           timeout = opts[:timeout] || 60
 
           nats = ::Protobuf::Nats.client_nats_connection
-          inbox = nats.new_inbox
 
           # Publish to server
-          sub = nats.subscribe(inbox, :max => 2)
-          nats.publish(subject, data, inbox)
-
-          # Wait for reply
-          first_message = nats.next_message(sub, ack_timeout)
-          return :ack_timeout if first_message.nil?
-          first_message_data = first_message.data
-          return :nack if first_message_data == ::Protobuf::Nats::Messages::NACK
-
-          second_message = nats.next_message(sub, timeout)
-          second_message_data = second_message.nil? ? nil : second_message.data
-
-          # Check messages
-          response = case ::Protobuf::Nats::Messages::ACK
-                     when first_message_data then second_message_data
-                     when second_message_data then first_message_data
-                     else return :ack_timeout
-                     end
-
-          fail(::Protobuf::Nats::Errors::ResponseTimeout, subject) unless response
-
-          response
-        ensure
-          # Ensure we don't leave a subscriptiosn sitting around.
-          nats.unsubscribe(sub)
+          with_subscription do |sub_inbox|
+            begin
+              completed_request = false
+
+              if !sub_inbox.subscription.is_valid # replace the subscription if is has been pooled but is no longer valid (maybe a reconnect)
```

Comment:
We should probably make `with_subscription` handle this logic?

Reply:
we could ... I didn't go so far as abstracting away getting a reference to the connection ... I don't know if issuing an `unsub` from another connection is valid